scripts/lldb/firestore.py

```diff
@@ -0,0 +1,250 @@
+
+    command script import ~/path/to/firebase-ios-sdk/scripts/lldb/firestore.py
+
+Most of this implementation is based on "Variable Formatting" in the LLDB online
+manual: https://lldb.llvm.org/use/variable.html. There are two major features
+we're making use of:
+
+  * Summary Providers: these are classes or functions that take an object and
+    produce a (typically one line) summary of the type
+
+  * Synthetic Children Providers: these are classes that provide an alternative
+    view of the data. The children that are synthesized here show up in the
+    graphical debugger.
+"""
+
+
+class ForwardingSynthProvider(object):
+  """A synthetic child provider that forwards all methods to another provider.
+
+  Override the `delegate` method to customize the target to which this forwards.
+  """
+
+  def __init__(self, value, params):
+    self.value = value
+
+  def delegate(self):
+    return self.value
+
+  def has_children(self):
+    return self.delegate().MightHaveChildren()
+
+  def num_children(self):
+    return self.delegate().GetNumChildren()
+
+  def get_child_index(self, name):
+    return self.delegate().GetIndexOfChildWithName(name)
+
+  def get_child_at_index(self, index):
+    return self.delegate().GetChildAtIndex(index)
+
+  def update(self):
+    pass
+
+
+# Abseil
+
+class AbseilOptional_SynthProvider(object):
+  """A synthetic child provider that hides the internals of absl::optional.
+  """
+
+  def __init__(self, value, params):
+    self.value = value
+    self.engaged = None
+    self.data = None
+
+  def update(self):
+    # Unwrap all the internal optional_data and similar types
+    value = self.value
+    while True:
+      if value.GetNumChildren() <= 0:
+        break
+
+      child = value.GetChildAtIndex(0)
+      if not child.IsValid():
+        break
+
+      if 'optional_internal' not in child.GetType().GetName():
+        break
+
+      value = child
+
+    # value should now point to the innermost absl::optional container type.
+    self.engaged = value.GetChildMemberWithName('engaged_')
+
+    if self.has_children():
+      self.data = value.GetChildMemberWithName('data_')
+
+    else:
+      self.data = None
+
+  def has_children(self):
+    return self.engaged.GetValueAsUnsigned(0) != 0
+
+  def num_children(self):
+    return 2 if self.has_children() else 1
+
+  def get_child_index(self, name):
+    if name == 'engaged_':
+      return 0
+    if name == 'data_':
+      return 1
+    return -1
+
+  def get_child_at_index(self, index):
+    if index == 0:
+      return self.engaged
+    if index == 1:
+      return self.data
+
+
+def AbseilOptional_SummaryProvider(value, params):
+  # Operates on the synthetic children above, calling has_children.
+  return 'engaged={0}'.format(format_bool(value.MightHaveChildren()))
+
+
+# model
+
+class DatabaseId_SynthProvider(ForwardingSynthProvider):
+  """Makes DatabaseId behave as if `*rep_` were inline."""
```

Comment:
Nit: perhaps add a little bit of context to make what `rep_` is more immediately obvious, e.g. `Hides the fact that DatabaseId uses PIMPL by making it behave as if...`?

Reply:
This is sort of veering into overkill. I don't really want to comment on the rationale for why something is laid out the way it is here. I did add some text to indicate that this hides the implementation details.